runtime/commonMain/src/kotlinx/serialization/SerialDescriptorBuilder.kt

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.internal.*
+
+/**
+ * Builder for [SerialDescriptor].
+ * The resulting descriptor will be uniquely identified by the given [serialName],
+ * with the corresponding [kind] and structure described in [builder] function.
+ * The count of descriptor elements should be known in advance to make API less error-prone,
+ * and this builder will throw [IllegalStateException] if count of added elements will be
+ * lesser or greater than [elementsCount].
+ *
+ * Example:
+ * ```
+ * // Class with custom serializer and custom serial descriptor
+ * class Data(
+ *     val intField: Int, // This field is ignored by custom serializer
+ *     val longField: Long, // This field is written as long, but in serialized form is named as "_longField"
+ *     val stringList: List<String> // This field is written as regular list of strings
+ *     val nullableInt: Int?
+ * )
+ * // Descriptor for such class:
+ * SerialDescriptor("my.package.Data", 3) {
+ *     // intField is deliberately ignored by serializer -- not present in the descriptor as well
+ *     element<Long>("_longField") // longField is named as _longField
+ *     element("stringField", listDescriptor<String>())
+ * }
+ * ```
+ */
+public fun SerialDescriptor(
+    serialName: String,
+    elementsCount: Int,
```

Comment:
I do not see why we should force users to pre-calculate this size if it is clearly can be deduced from `element` calls count